cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -398,6 +403,93 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 	return nil
 }
 
+func (a *StaticAutoscaler) handleQuoteAndStockoutErrors(currentTime time.Time) {
+	cloudProviderNodeInstances := a.clusterStateRegistry.GetCloudProviderNodeInstances()
+
+	nodeGroups := a.CloudProvider.NodeGroups()
+	for _, nodeGroup := range nodeGroups {
+		instances, found := cloudProviderNodeInstances[nodeGroup.Id()]
+		if !found {
+			continue
+		}
+
+		outOfResourcesInstancesByErrorCode := make(map[string][]cloudprovider.Instance)
+		for _, instance := range instances {
+			if instance.Status != nil && instance.Status.State == cloudprovider.StateCreating && instance.Status.ErrorInfo != nil {
+				errorInfo := instance.Status.ErrorInfo
+				if errorInfo.ErrorClass == cloudprovider.ErrorOutOfResources {
+					errorInstances, found := outOfResourcesInstancesByErrorCode[errorInfo.ErrorCode]
+					if !found {
+						errorInstances = make([]cloudprovider.Instance, 0, 0)
+					}
+					errorInstances = append(errorInstances, instance)
+					outOfResourcesInstancesByErrorCode[errorInfo.ErrorCode] = errorInstances
+				}
+			}
+		}
+
+		for errorCode, instances := range outOfResourcesInstancesByErrorCode {
+			klog.Infof("Failed adding %v nodes to group %v due to %v", len(instances), nodeGroup.Id(), errorCode)
+			a.LogRecorder.Eventf(
+				apiv1.EventTypeWarning,
+				"ScaleUpFailed",
+				"Failed adding %v nodes to group %v due to %v",
+				len(instances),
+				nodeGroup.Id(),
+				errorCode)
+
+			instanceNames := make([]string, 0, 0)
+			for _, instance := range instances {
+				instanceNames = append(instanceNames, instance.Id)
+			}
+
+			if a.deleteErroredNodesBeingCreated(nodeGroup, instanceNames, currentTime) {
+				a.clusterStateRegistry.RegisterFailedScaleUp(nodeGroup, metrics.FailedScaleUpReason(errorCode), currentTime)
+			}
+		}
+	}
+}
+
+func (a *StaticAutoscaler) deleteErroredNodesBeingCreated(nodeGroup cloudprovider.NodeGroup, nodeNames []string, currentTime time.Time) bool {
+	nodes := make([]*apiv1.Node, 0, len(nodeNames))
+	for _, nodeName := range nodeNames {
+		if a.clusterStateRegistry.WasDeleteNodeRequested(nodeName) {
+			continue
+		}
+		// create dummy node name
+		node := &apiv1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nodeName,
+			},
+			Spec: apiv1.NodeSpec{
+				ProviderID: nodeName,
+			},
+		}
+		nodes = append(nodes, node)
+	}
+
+	if len(nodes) == 0 {
+		return false
+	}
+
+	// Issue a delete
+	err := nodeGroup.DeleteNodes(nodes)
+	if err != nil {
+		klog.Warningf("Error while trying to delete nodes from %v: %v", nodeGroup.Id(), err)
+		return false
+	}
+
+	// Remember the fact that we deleted the nodes
+	// Question: Maybe for safety we should mark it in the loop above.
```

Comment:
That's an interesting problem. I see what you're doing here - only on successful delete call to cloud provider we'll update failed scale ups metric and back-off, which will hopefully mean we do it just once... unless we restart. But if the instance is stuck and can't be deleted, it means we'll never back-off the node group at all. On the other hand, we'll keep publishing events about it every loop.

A side-effect is also that if the instance delete fails, we'll keep retrying every loop, which may not be the best idea if there's a problem with the underlying service which may be limited to a specific zone.

I'm not sure if it's worth the added complexity to attempt to solve this 100% correctly, but perhaps we can consider the priorities to make the right trade-offs?

The way I see, from most to least important:
- don't get stuck trying to delete the same instance every loop if it doesn't work,
- back-off the group ASAP,
- don't update the metric twice if possible (loss of information is somehow more acceptable, since it'll be the case on restarts anyway - but let's not do false positives),
- everything else (events, metric loss).

I wonder if something like this may work:
- as soon as errors are recognized, report instance ids to cluster state and save,
- on first seeing a given instance id, back-off the node group,
- on first seeing a given instance id, update the metric but only if there's a scale up in progress,
- delete instances in a way that won't block us (i.e. if it fails don't retry in every iteration, scale up other groups in the meantime).

Still figuring out how to fit updating scale up information in it. My initial thought was that we should also do it ASAP (when updating metric), but then we'd have to triple-check how it affects everything else.

Reply:
OK, so it seems we ignore that error and don't fail the loop, so we'll keep scaling the rest of the cluster no matter what (which is good).

I still think we should back-off even if we don't manage to delete the node for whatever reason (e.g. because the group is at or below minimum), and publish metric if there was an active scale-up attempt.

Reply:
I think the logic with saving last-seen-instances in CSR can work for this. And I really think we should only do failed scale up stuff (metrics, events, etc.) IF there's a scale up in progress in that group, and update that scale up soon-ish.

That way, if a user resizes group, but we don't know about it, we'll not start reporting failed scale up. And on restart, we'll also not do that (because we lost scale up state).

Reply:
Updated